### ISSUE TITLE
Horovod crashed with Auto Mixed Precision Training

### ISSUE THREAD

**Issue by garudareiga** [horovod/horovod#1738]
**Environment:**
1. Framework: TensorFlow
2. Framework version: 1.14
3. Horovod version: 0.18
4. MPI version: 4.0.1
5. CUDA version:
6. NCCL version:
7. Python version:
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Bug report:**
=== Context ===

I’m experimenting with auto mixed precision training in TensorFlow, and we use DynamicLossScale to scale up the loss value to compensate the precision loss. Therefore, the grads will be scaled and unscaled before we do AllReduce in Horovod.

=== Problem ===
Currently, Horovod crashed for 2 of our models when it’s doing AllReduce. It’s crashed after running for 10 hours or 20 hours.
```
File "/usr/local/lib/python2.7/dist-packages/horovod/tensorflow/__init__.py", line 80, in allreduce summed_tensor_compressed = _allreduce(tensor_compressed)
File "/usr/local/lib/python2.7/dist-packages/horovod/tensorflow/mpi_ops.py", line 86, in _allreduce return MPI_LIB.horovod_allreduce(tensor, name=name) 
File "<string>", line 80, in horovod_allreduce 
File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/op_def_library.py", line 788, in _apply_op_helper op_def=op_def) 
File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/util/deprecation.py", line 507, in new_func return func(*args, **kwargs) 
File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 3616, in create_op op_def=op_def) 
File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 2005, in __init__ self._traceback = tf_stack.extract_stack()
```

DynamicLossScale will gradually increase the LossScale multiplier until grads overflows and becomes infinite, and then decrease. How Horovod MPI do "sum" of tensors if one worker produces these infinite gradients and other workers do not? The default behavior in TensorFlow is that it will detect if grad is infinite. If so, it will skip applying the gradients, but it’s happening after Hovrovod AllReduce.

**Comment by tgaddair** [comment]
Hey @garudareiga, thanks for raising this issue.  Can you share the full stack trace and a simple example we can use to repro?

@romerojosh is looking into ways of addressing issues of overflow / underflow of the allreduce when using mixed precision.  Possibly the solution he's working on would be relevant to this as well.

**Comment by romerojosh** [comment]
Automatic mixed precision should work with Horovod correctly. The allreduce occurs _before_ dynamic loss scaling and will propagate any infinite gradient values if they occur on any worker (i.e. summing infinite or NaN values is not problem in MPI or NCCL). Then dynamic loss scaling is applied based on the allreduced gradients so it acts consistently across all workers. 

Can you share how you are enabling automatic mixed precision in your script?
